### PR TITLE
Feature/admin api skeleton

### PR DESCRIPTION
**Pull request by axelsiebert**
## Changes

- Add API endpoint stups for compositions, contributions, ehr, directories and templates
- Activate Admin API via environment variables or properties files
- Activate delete all resources via environment variable or properties files

## Related issue

Closes https://github.com/ehrbase/project_management/issues/299

## Additional information and checks

- [x] Pull request linked in changelog

### CODE REVIEW

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+                            )
+                    }
+            ),
+            @ApiResponse(
+                    code = 401,
+                    message = "Client credentials are invalid or have been expired."
+            ),
+            @ApiResponse(
+                    code = 403,
+                    message = "Client has not permission to access this resource since admin role is missing."
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java:71]: Have you tested this? Are those messages the ones in the actual response? I would think Spring Boot captures and denies the request earlier than those lines are "executed". So if that's the case, those lines could be removed and handled in a more general place (or just left with the default).

This also applies to basically all other admin endpoints.

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        // Check if EHR exists
+        if (!ehrService.hasEhr(ehrUuid)) {
+            throw new ObjectNotFoundException(
+                    "Admin Composition", String.format("EHR with id %s does not exist.", ehrId)
+            );
+        }
+
+        // Composition existence will be checked during deletion
+
+        return ResponseEntity.ok().body(new AdminDeleteResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java:97]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminContributionController.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        // Check if EHR exists
+        if (!this.ehrService.hasEhr(ehrUuid)) {
+            throw new ObjectNotFoundException(
+                    "Admin Contribution", String.format("EHR with id %s does not exist", ehrId)
+            );
+        }
+
+        // Contribution existence check will be done in services
+
+        return ResponseEntity.ok().body(new AdminUpdateResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminContributionController.java:96]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminContributionController.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        // Check if EHR exists
+        if (!this.ehrService.hasEhr(ehrUuid)) {
+            throw new ObjectNotFoundException(
+                    "Admin Contribution", String.format("EHR with id %s does not exist", ehrId)
+            );
+        }
+
+        // Contribution existence check will be done in services
+
+        return ResponseEntity.ok().body(new AdminDeleteResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminContributionController.java:144]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminDirectoryController.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        // Check if EHR exists
+        if (!this.ehrService.hasEhr(ehrUuid)) {
+            throw new ObjectNotFoundException(
+                    "Admin Directory", String.format("EHR with id %s does not exist", ehrId)
+            );
+        }
+
+        // Directory existence check will be done during processing
+
+        return ResponseEntity.ok().body(new AdminDeleteResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminDirectoryController.java:98]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminEhrController.java [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                    String ehrId
+    ) {
+
+        // Check if EHR with id exists
+        UUID ehrUuid = UUID.fromString(ehrId);
+        if (!ehrService.hasEhr(ehrUuid)) {
+            throw new ObjectNotFoundException("Admin EHR", String.format("EHR with id %s does not exist.", ehrId));
+        }
+
+        return ResponseEntity.ok().body(new AdminUpdateResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminEhrController.java:93]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminEhrController.java [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                    String ehrId
+    ) {
+
+        UUID ehrUuid = UUID.fromString(ehrId);
+        // Check if EHR with id exists
+        if (!ehrService.hasEhr(ehrUuid)) {
+            throw new ObjectNotFoundException("Admin EHR", String.format("EHR with id %s does not exist.", ehrId));
+        }
+
+        return ResponseEntity.ok().body(new AdminDeleteResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminEhrController.java:134]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminTemplateController.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+                    message = "Template could not be found."
+            )
+    })
+    public ResponseEntity<AdminUpdateResponseData> updateTemplate(
+            @ApiParam(value = "Target template id to update")
+            @PathVariable(value = "template_id")
+                    String templateId
+    ) {
+
+        return ResponseEntity.ok().body(new AdminUpdateResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminTemplateController.java:76]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminTemplateController.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+                    message = "Template could not be found."
+            )
+    })
+    public ResponseEntity<AdminDeleteResponseData> deleteTemplate(
+            @ApiParam(value = "Target template id to update")
+            @PathVariable(value = "template_id")
+                    String templateId
+    ) {
+
+        return ResponseEntity.ok().body(new AdminDeleteResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminTemplateController.java:111]: So this is TODO, right? Can you add a TODO comment?

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminTemplateController.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+    public ResponseEntity<?> deleteAllTemplates() {
+
+        if (!this.adminApiConfiguration.getAllowDeleteAll()) {
+            return ResponseEntity
+                    .status(HttpStatus.METHOD_NOT_ALLOWED)
+                    .body(new AdminStatusResponseData(
+                            "Delete all resources not allowed."
+                    ));
+        }
+        return ResponseEntity.ok().body(new AdminDeleteResponseData(0));
```
Comment (jakesmolka) [inline rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminTemplateController.java:149]: So this is TODO, right? Can you add a TODO comment?

File: application/src/main/resources/application.yml [hunk 11]
```diff
@@ -45,6 +45,10 @@ cache:
 system:
   allow-template-overwrite: false
 
+admin-api:
+  active: ${ADMIN_API_ACTIVE:false}
+  allowDeleteAll: ${ADMIN_API_ALLOW_DELETE_ALL:false}
```
[the remark this answers is not in the document]
Comment (axelsiebert) [reply application/src/main/resources/application.yml:50]: I think it is the best way to do it in the defaults way. The point here is that I could not found a concrete documentation how to adress various config variables with dashes as you mentioned. 

Handling these values in my way the developer could always look into the file to see which environment variables are supported and which default value is configured for each.

File: application/src/main/resources/application-cloud.yml [hunk 12]
```diff
@@ -24,3 +24,7 @@ server:
 
 system:
   type: LOCAL
+
+admin-api:
+  active: ${ADMIN_API_ACTIVE:false}
+  allowDeleteAll: ${ADMIN_API_ALLOW_DELETE_ALL:false}
```
[the remark this answers is not in the document]
Comment (axelsiebert) [reply application/src/main/resources/application-cloud.yml:30]: Okay, removed from all files except application.yml

File: CHANGELOG.md [hunk 13]
```diff
@@ -16,7 +16,9 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ## [0.13.0] (beta)
 
 ### Added
-- added support for various functions in AQL (aggregation, statistical, string etc.) (see: https://github.com/ehrbase/ehrbase/pull/223/)
+
+- Added support for various functions in AQL (aggregation, statistical, string etc.) (see: https://github.com/ehrbase/ehrbase/pull/223/)
+- Add admin API endpoint stubs (see: https://github.com/ehrbase/ehrbase/pull/280)
```
[the remark this answers is not in the document]
Comment (axelsiebert) [reply CHANGELOG.md:21]: Okay, that was by accident ;-)

File: rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java (hunk 1 above)
Comment (axelsiebert) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java:71]: I thought these annotations are only for swagger documentation. 

You are right the handling of these cases is done by spring boot

Comment (jakesmolka) [reply rest-openehr/src/main/java/org/ehrbase/rest/openehr/controller/admin/OpenehrAdminCompositionController.java:71]: Oh, okay, if those are only the swagger ones everything is fine :)

**Pull request merged**